activemodel/lib/active_model/type/model.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  module Type
+    class Model < Value # :nodoc:
+      def initialize(**args)
+        @class_name = args.delete(:class_name)
+        @serializer = args.delete(:serializer) || ActiveSupport::JSON
+        super
+      end
+
+      def changed_in_place?(raw_old_value, value)
+        old_value = deserialize(raw_old_value)
+        old_value.attributes != value.attributes
+      end
+
+      def valid_value?(value)
+        return valid_hash?(value) if value.is_a?(Hash)
+
+        value.is_a?(klass)
+      end
+
+      def type
+        :model
+      end
+
+      def serializable?(value)
+        value.is_a?(klass)
+      end
+
+      def serialize(value)
+        serializer.encode(value.attributes_for_database)
+      end
+
+      def deserialize(value)
+        attributes = serializer.decode(value)
+        klass.new(attributes)
+      end
+
+      private
+        attr_reader :serializer
+
+        def valid_hash?(value)
+          value.keys.map(&:to_s).difference(klass.attribute_names).none?
```

Comment:
This implementation doesn't accommodate aliased attributes (defined with [alias_attribute](https://edgeapi.rubyonrails.org/classes/ActiveModel/AttributeMethods/ClassMethods.html#method-i-alias_attribute)).

```suggestion
          attribute_keys = value.keys.map do |key|
            key = key.to_s
            
            klass.attribute_alias?(key) ? klass.attribute_alias(key) : key
          end
          
          attribute_keys.difference(klass.attribute_names).none?
```

Reply:
Reflecting further, is it crucial for `:model` types to reject `Hash` arguments on any grounds? Couldn't that be specific to each implementation? I'm imagining an `ActiveModel::Model` (or any `Object`, really) that relies on `method_missing` or overrides `attributes=(...)` or `assign_attributes(...)` in a way that transforms attributes being passed in as part of assignment.

For example, what about a model that can assign `camelCased` JSON keys, but transform them into `snake_case` keys to support idiomatic Ruby methods?